### PR TITLE
Add training and test functions to integrate the native XGBoost library

### PR DESCRIPTION
**Pull request by maropu**
I'm working on the integration of the `XGBoost`.
As a first step, each UDTF training worker simply outputs a single `xgboost` model, and then a testing phase computes prediction values for each built model and averages them.
```
-- `xgboost_model` outputs (string model_id, byte[] model) where each UDTF process
-- creates a xgboost model with an unique ID.
CREATE TABLE xgboost_models AS
  SELECT train_xgboost(add_bias("features"), "label")
    FROM training_data;

-- prediction
CREATE TABLE test_data_with_id AS
  SELECT rowid(), features
    FROM test_data

SELECT rowid, AVG(predicted)
  FROM (
      SELECT xgboost_predict(rowid, features, model_id, model) AS (rowid, predicted)
        FROM xgboost_models CROSS JOIN test_data_with_id
    )
  GROUP BY rowid
```

### CODE REVIEW

File: core/pom.xml [hunk 1]
```diff
@@ -90,6 +90,13 @@
 			<version>1.0.3</version>
 			<scope>compile</scope>
 		</dependency>
+		<dependency>
+			<groupId>ml.dmlc</groupId>
+			<artifactId>xgboost4j</artifactId>
+			<version>0.5</version>
+			<scope>system</scope>
+			<systemPath>${basedir}/lib/xgboost4j-0.5-jar-with-dependencies.jar</systemPath>
+		</dependency>
```
Comment (maropu) [inline core/pom.xml]: Currently, no `xgboost4j` exists in the maven repository. I'll ask the author later why the library is not uploaded there.

**Comment by maropu** [comment]
An obvious other problem is that the native `xgboost` library depends on some shared libraries;
```
$ otool -L libxgboost4j.dylib 
libxgboost4j.dylib:
        jvm-packages/lib/libxgboost4j.so (compatibility version 0.0.0, current version 0.0.0)
        /usr/local/lib/gcc/4.9/libstdc++.6.dylib (compatibility version 7.0.0, current version 7.20.0)
        /usr/local/lib/gcc/4.9/libgomp.1.dylib (compatibility version 2.0.0, current version 2.0.0)
        /usr/lib/libSystem.B.dylib (compatibility version 1.0.0, current version 1213.0.0)
        /usr/local/lib/gcc/4.9/libgcc_s.1.dylib (compatibility version 1.0.0, current version 1.0.0)
``` 
Also, `xgboost4j` includes no native library for each platform by default. That is, you need compile it for your own platform. This could make it harder to deploy `xgboost4j` in a large heterogenous computing cluster. To mitigate the portability and deployment issues, we might need to fix these issues directly in the original `xgboost4j` codes. Otherwise, we need to fork and fix them.

**Comment by maropu** [comment]
I think `snappy-java` is a winner example in these kinda issues.

**Comment by maropu** [comment]
I asked `xgboost4j` guys in [here](https://github.com/dmlc/xgboost/issues/935).

**Comment by myui** [comment]
better to create our own package `xgboost4j-native` including native libraries if they are not positive to make it.

**Comment by maropu** [comment]
okay

File: core/src/main/java/hivemall/utils/hadoop/HiveUtils.java [hunk 2]
```diff
@@ -359,6 +359,21 @@ public static double getAsConstDouble(@Nonnull final ObjectInspector numberOI)
                 + TypeInfoUtils.getTypeInfoFromObjectInspector(numberOI));
     }
 
+    @SuppressWarnings("unchecked")
```
Comment (myui) [inline core/src/main/java/hivemall/utils/hadoop/HiveUtils.java]: Please move to Primitives, not related to Hive.

Comment (maropu) [inline core/src/main/java/hivemall/utils/hadoop/HiveUtils.java]: okay, I'll fix it.

File: core/src/main/java/hivemall/xgboost/regression/XgBoostUDTF.java [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+            final DMatrix trainData = new DMatrix(featuresList.iterator(), "");
+
+            // Kick off training with xgboost
+            final Booster booster = createXgBooster(xgboostParams, trainData);
+            for(int i = 0; i < num_boosting_iter; i++) {
+                booster.update(trainData, i);
+            }
+
+            // Output the built model
+            final String modelId = generateUniqueIdentifier();
```
Comment (myui) [inline core/src/main/java/hivemall/xgboost/regression/XgBoostUDTF.java]: You can use `HadoopUtils.getTaskId()`.

**Comment by maropu** [comment]
I implemented a XGBoost multiclass classifier;
```
-- `xgboost_model` outputs (string model_id, byte[] model) where each UDTF process
-- creates a xgboost model with an unique ID.
CREATE TABLE xgboost_models AS
  SELECT train_multiclass_xgboost_classifier(add_bias("features"), "label")
    FROM training_data;

-- prediction
CREATE TABLE test_data_with_id AS
  SELECT rowid(), features
    FROM test_data

SELECT rowid, max_row(avg_prob_per_label, label)
  FROM (
      SELECT rowid, label, AVG(probability) AS avg_prob_per_label
        FROM (
            SELECT xgboost_multiclass_predict(rowid, features, model_id, model) AS (rowid, label, probability)
              FROM xgboost_models CROSS JOIN test_data_with_id
          )
        GROUP BY rowid, label
    )
  GROUP BY rowid
```

**Comment by maropu** [comment]
I did a XGBoost binary classifier;
```
-- `xgboost_model` outputs (string model_id, byte[] model) where each UDTF process
-- creates a xgboost model with an unique ID.
CREATE TABLE xgboost_models AS
  SELECT train_xgboost_classifier(add_bias("features"), "label")
    FROM training_data;

-- prediction
CREATE TABLE test_data_with_id AS
  SELECT rowid(), features
    FROM test_data

SELECT rowid, CAST((case when AVG(predicted) >= 0.5 then 1.0 else 0.0 end) as FLOAT) as predicted
  FROM (
      SELECT xgboost_predict(rowid, features, model_id, model) AS (rowid, predicted)
        FROM xgboost_models CROSS JOIN test_data_with_id
    )
  GROUP BY rowid
```

**Comment by myui** [comment]
ref. to discussions https://github.com/myui/hivemall/issues/251

**Comment by maropu** [comment]
How to use `XGBoost` functions on DataFrame/Spark is as follows;
```
// Load libsvm-formatted training data 
val trainDf = sqlContext.sparkSession.read.format("libsvm").load("E2006.train")

// Load test data
val testDf = sqlContext.sparkSession.read.format("libsvm").load("E2006.test")
  .withColumn("rowid", rowid()).cache

// Set XGBoost options here
val xgbOptions = XGBoostOptions()
  .set("num_round", "10000")
  .set("max_depth", "32")

// Build models with the XGBoost and write them into persistent storages
trainDf.train_xgboost_regr($"features", $"label", s"${xgbOptions}")
  .write.format(xgboost).load("xgboost_models")

// Load models from the storage
val model = sqlContext.sparkSession.read.format(xgboost).load("xgboost_models")

// Do prediction
val predict = model.join(testDf).xgboost_predict($"rowid", $"features", $"model_id", $"pred_model")
  .groupBy("rowid").avg()
  .as("rowid", "predicted")

// Check precision
val result = predict.join(testDf, predict("rowid") === testDf("rowid"), "INNER")
result.select(avg(abs($"predicted" - $"target"))).show
```

File: xgboost/src/main/java/hivemall/xgboost/XGBoostUtils.java [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ * limitations under the License.
+ */
+package hivemall.xgboost;
+
+import ml.dmlc.xgboost4j.LabeledPoint;
+
+import java.util.List;
+
+public final class XGBoostUtils {
+
```
Comment (myui) [inline xgboost/src/main/java/hivemall/xgboost/XGBoostUtils.java]: Could you put `private XGBoostUtils() {}` to avoid instantiation?

Comment (maropu) [inline xgboost/src/main/java/hivemall/xgboost/XGBoostUtils.java]: fixed

**Comment by maropu** [comment]
To easily use `XGBoost` in major platforms, I put two pre-comipled binaries for `Linux/x86_64` and `Mac/x86_64` in `xgboost4j-with-dependeicies.jar` The binary for  `Linux/x86_64` was compiled with `g++-4.6` because of [the min version of `g++` that `XGBoost` supports](https://github.com/dmlc/xgboost/blob/master/doc/build.md#build-the-shared-library). The other binary was compiled with `g++-4.7` because I couldn't compile it with `g++-4.6` in my Mac/x86_64 platform. Both binaries were statically linked with glibc++ and this is a not a license violation because of [the linking exception](https://en.wikipedia.org/wiki/GPL_linking_exception).

**Comment by maropu** [comment]
Also, I turned off `OpenMP` in both binaries.

File: .travis.yml [hunk 5]
```diff
@@ -36,8 +36,9 @@ notifications:
   email: false
 
 script:
-  - mvn -q test -Pspark-1.6
   - mvn -q test -Pspark-2.0
+  # test the spark-1.6 module only in this second run
+  - mvn -q test -Pspark-1.6 -Dtest=org.apache.spark.*
```
Comment (myui) [inline .travis.yml]: I think it should be tested both on spark 1.6 and spark 2.0.
Before committing this change, test was run successfully. Is this change required?

**Comment by myui** [comment]
@maropu I'm concern about LGPL license issue. It would be a problem in the Apache Incubation process.

LZO is not bundled in Apache Hadoop due to licensing issues.

And we SHOULD not include LGPL library in our release.
It's okey to provide a build option to statically link GLIBC though.

File: .travis.yml (hunk 5 above)
Comment (maropu) [inline .travis.yml]: yea, I think so.
IIUC the first tests the `Hivemall` core stuffs (e.g., core, nlp, and mixserv) and the spark-2.0 module.
The other tests the spark-1.6 module only because the `Hivemall` core stuff has already been tested in the first test.

**Comment by maropu** [comment]
This is not the same case with LZE one. Anyway, linking with `libstdc++`(LGPL) is not a problem and see the similar discussion: https://github.com/xerial/snappy-java/issues/113. Both binaries include a small part of `libstdc++` by the static links though, they do not distribute `libstdc++` as an independent library (See: https://www.gnu.org/licenses/gcc-exception-3.1-faq.en.html).
Thought?

**Comment by myui** [comment]
We should not follow Snappy-java way because the discussion by IP review is unclear while the github issue is closed.

GCC exception cause is another issue to glibc. Does glibc has exception cause?

**Comment by myui** [comment]
http://libcxx.llvm.org/
http://www.wdic.org/w/TECH/libc%2B%2B

**Comment by maropu** [comment]
okay, I removed the dependencies (`libgcc` and `libstdc++`).
Could you check this again?

File: spark/spark-1.6/pom.xml [hunk 6]
```diff
@@ -9,13 +9,12 @@
 		<relativePath>../../pom.xml</relativePath>
 	</parent>
 
-	<artifactId>hivemall-spark</artifactId>
-	<name>Hivemall on Spark</name>
+	<artifactId>hivemall-spark_2.11</artifactId>
```
Comment (myui) [inline spark/spark-1.6/pom.xml]: why `2.11?` hivemall-spark-v1 and hivemall-spark-v2 might be good for artifact id.

File: spark/spark-2.0/pom.xml [hunk 7]
```diff
@@ -9,14 +9,13 @@
 		<relativePath>../../pom.xml</relativePath>
 	</parent>
 
-	<artifactId>hivemall-spark</artifactId>
-	<name>Hivemall on Spark</name>
+	<artifactId>hivemall-spark_2.11</artifactId>
```
Comment (myui) [inline spark/spark-2.0/pom.xml]: Better to use `<project><properties>` for version string.

File: spark/spark-1.6/pom.xml (hunk 6 above)
Comment (maropu) [inline spark/spark-1.6/pom.xml]: This means this jar supports spark-2.0 compiled with scala-2.11.
This naming rule is along with sprak jar.

File: spark/spark-2.0/pom.xml (hunk 7 above)
Comment (maropu) [inline spark/spark-2.0/pom.xml]: you mean like this?
```
<properties><scala.version>2.11</scala.version></properties>
...
<artifactId>hivemall-spark_${scala.version}</artifactId>
```

Comment (myui) [inline spark/spark-2.0/pom.xml]: yes, you can use properties in the top-level pom.xml for cross-modules global properties.

Comment (maropu) [inline spark/spark-2.0/pom.xml]: okay

Comment (maropu) [inline spark/spark-2.0/pom.xml]: done

**Comment by maropu** [comment]
fyi: If you get some errors in a `xgboost` binary bundled in the jar, you can compile the binary on your platform: `mvn -Pcompile-xgboost clean package`

**Comment by myui** [comment]
@maropu LGTM. Merged. Well done.
